### PR TITLE
withDistributedBatches refactor

### PR DESCRIPTION
**Pull request by ddl-giuliocapolino**
### Description
For performance and ease-of-use purposes, users may need to access data in Arrow.Record format directly from queries, rather than getting it in sql.Rows format and having to covert it manually. This can be achieved via a flag in Go's context, which is consistent with other workflows in gosnowflake. This PR covers the impl to get these records via a new API endpoint: gosnowflake writes arrays of records to an array of ResultBatch objects, which can be retrieved by casting the query rows object into a SnowflakeRows object, and calling the `GetBatches` endpoint.

Furthermore, the data retrieved from Snowflake is stored in arrow.Record objects. The data stored in these objects, however, is not in a format that can be sent to a client. In order to transform it to its final format, we add a converter along with tests. The conversion documentation can be found [here](https://docs.google.com/document/d/1giKZmbMf5mnxTTnNL3-L4CQXhV2Zl_FwTAq64qeesTo/edit?usp=sharing).
### Checklist
- [ ] Code compiles correctly
- [ ] Run ``make fmt`` to fix inconsistent formats
- [ ] Run ``make lint`` to get lint errors and fix all of them
- [ ] Created tests which fail without the change (if possible)
- [ ] All tests passing
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: rows.go [hunk 1]
```diff
@@ -27,7 +27,7 @@ var (
 	maxChunkDownloaderErrorCounter = 5
 )
 
-type snowflakeRows struct {
+type SnowflakeRows struct {
```
Comment (ddl-giuliocapolino) [inline rows.go]: before I add comments to all of the following, I'd like to confirm this is the one and only way we can pursue for this (ie turning `snowflakeRows` into `SnowflakeRows`):

Screen Shot 2022-01-19 at 1 03 27 AM

File: chunk_downloader.go [hunk 2]
```diff
@@ -664,3 +701,32 @@ func copyChunkStream(body io.Reader, rows chan<- []*string) error {
...
+// Fetch returns an array of records representing a chunk in the query
+func (rb *ResultBatch) Fetch() ([]array.Record, error) {
+	if len(rb.scd.ResultBatches) == 1 {
+		return rb.scd.ResultBatches[0].rec, nil
+	}
+	err := rb.funcDownloadHelper(context.Background(), rb.scd, rb.idx)
+	if err != nil {
+		return nil, err
+	}
+	return rb.scd.ResultBatches[rb.idx].rec, nil
```
Comment (ddl-giuliocapolino) [inline chunk_downloader.go]: perhaps fetch should only return `error` since we're putting the `[]array.Record` inside the `.rec` field already

Comment (ddl-giuliocapolino) [reply chunk_downloader.go]: in that case, we'd need to turn `.rec` into `.Rec`. Thoughts @ddl-gabrielhaim ?

Comment (ddl-giuliocapolino) [reply chunk_downloader.go]: rationale: we probably don't want to store the full record both on the `ResultBatch` object AND upon returning it and thus saving it in memory into another object. Unless we want to do both and leave it up to the client to decide - I like the return `error` and exporting `Rec` personally.

File: rows.go (hunk 1 above)
Comment (ddl-giuliocapolino) [reply rows.go]: not applicable anymore with new commit (amended).

File: chunk_downloader.go (hunk 2 above)
Comment (ddl-giuliocapolino) [reply chunk_downloader.go]: I exposed `.Rec` in new commit (amended). These msgs are not applicable anymore.

File: chunk_arrow.go [hunk 3]
```diff
@@ -60,57 +52,39 @@ func (arc *arrowResultChunk) decodeArrowChunk(ctx context.Context, rowType []exe
...
+			return nil, err
 		}
-		record, err := arrowToRecord(rawRecord, rowType)
+		record, err := arrowToRecord(rawRecord, scd.RowSet.RowType)
 		rawRecord.Release()
 		if err != nil {
-			return numRows, err
+			return nil, err
 		}
+		record.Retain()
```
Comment (ddl-gabrielhaim) [inline chunk_arrow.go:70]: isn't it strange to retain this?

File: chunk_downloader.go [hunk 4]
```diff
@@ -90,23 +92,22 @@ func (scd *snowflakeChunkDownloader) nextResultSet() error {
 }
 
 func (scd *snowflakeChunkDownloader) start() error {
+	if usesDistributedBatches(scd.ctx) {
+		return setUpDistributedBatches(scd)
```
Comment (ddl-gabrielhaim) [inline chunk_downloader.go]: maybe use a similar name? `startDistributedBatches`?

File: chunk_downloader.go [hunk 5]
```diff
@@ -664,3 +684,32 @@ func copyChunkStream(body io.Reader, rows chan<- []*string) error {
 	}
 	return nil
 }
+
+// A ResultBatch object encapsulates a function that retrieves a subset of rows in a result set.
```
Comment (ddl-gabrielhaim) [inline chunk_downloader.go]: nit: godoc starts with the name of the object you are describing (remove `A`)

File: chunk_downloader.go [hunk 6]
```diff
@@ -664,3 +684,32 @@ func copyChunkStream(body io.Reader, rows chan<- []*string) error {
...
+// Fetch returns an array of records representing a chunk in the query
+func (rb *ResultBatch) Fetch() error {
+	if rb.Rec != nil {
+		return nil
+	}
+	err := rb.funcDownloadHelper(context.Background(), rb.scd, rb.idx)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (ddl-gabrielhaim) [inline chunk_downloader.go]: no need to test `err` just `return rb.funcDownloadHelper(context.Background(), rb.scd, rb.idx)`

File: chunk_downloader.go [hunk 7]
```diff
@@ -664,3 +684,32 @@ func copyChunkStream(body io.Reader, rows chan<- []*string) error {
...
+}
+
+func usesDistributedBatches(ctx context.Context) bool {
+	val := ctx.Value(distributedResultBatches)
+	if val == nil {
+		return false
+	}
+	a, ok := val.(bool)
+	return ok && a
```
Comment (ddl-gabrielhaim) [inline chunk_downloader.go]: you had a chance for a-ok (https://en.wikipedia.org/wiki/A-okay) and you missed it

File: chunk_downloader.go (hunk 6 above)
Comment (ddl-giuliocapolino) [reply chunk_downloader.go]: whops thanks!

File: chunk_downloader.go (hunk 7 above)
Comment (ddl-giuliocapolino) [reply chunk_downloader.go]: I am ashamed of myself

File: chunk_arrow.go (hunk 3 above)
Comment (ddl-giuliocapolino) [reply chunk_arrow.go:70]: get a `malformed data` in the do_get otherwise :(

File: chunk_downloader.go [hunk 8]
```diff
@@ -90,23 +92,22 @@ func (scd *snowflakeChunkDownloader) nextResultSet() error {
 }
 
 func (scd *snowflakeChunkDownloader) start() error {
+	if usesDistributedBatches(scd.ctx) {
+		return startDistributedBatches(scd)
```
Comment (ddl-gabrielhaim) [inline chunk_downloader.go]: why aren't you making this a method of `sdc`?

File: chunk_downloader.go [hunk 9]
```diff
@@ -265,6 +261,33 @@ func getChunk(
 	return newRetryHTTP(ctx, scd.sc.rest.Client, http.NewRequest, u, headers, timeout).execute()
 }
 
+func startDistributedBatches(scd *snowflakeChunkDownloader) error {
```
Comment (ddl-gabrielhaim) [inline chunk_downloader.go]: re: previous comment `func (sdc *snowflakeChunkDownloader) startDistributedBatches() error {...}`

File: chunk_downloader.go [hunk 10]
```diff
@@ -400,17 +423,15 @@ func decodeChunk(scd *snowflakeChunkDownloader, idx int, bufStream *bufio.Reader
...
-		var ctx context.Context
-		if isArrowRecordChanEnabled(scd.ctx) {
-			ch := make(chan []array.Record, 1)
-			scd.ArrowChannels.Store(idx, ch)
-			ctx = context.WithValue(scd.ctx, arrowRecordChannel, ch)
-		} else {
-			ctx = scd.ctx
+		if usesDistributedBatches(scd.ctx) {
+			scd.ResultBatches[idx].Rec, err = arc.decodeArrowBatch(scd)
+			if err != nil {
```
Comment (ddl-gabrielhaim) [inline chunk_downloader.go]: nit: should be able to combine those two statements

**Review by ddl-gabrielhaim — APPROVED** [review]

**Pull request merged**